### PR TITLE
PR Chapter 01 - Writing your first module

### PR DESCRIPTION
**Pull request by manuakasam**
Discussion about the first chapter goes here :)

### CODE REVIEW

File: 01_writing_first_module.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+class Module implements ConfigProviderInterface
+{
+    public function getConfig()
+    {
+        return array();
+    }
+}
+```
+
+With this we let the [`ModuleManager`](https://github.com/zendframework/zf2/:current_branch/library/Zend/ModuleManager/ModuleManager.php)
```
Comment (tfountain) [inline 01_writing_first_module.md]: I think this section needs rewording now as some of this info was incorporated in my edit in the paragraph above. How about just:

> To help keep our project organized we're going to put our array configuration in a separate file. Go ahead and create this file at `/module/Album/config/module.config.php`:

File: 01_writing_first_module.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+            )
+        )
+    )
+);
+```
+
+With this we have created a new route called `album-default` that listens to the url `domain.loc/album`. Whenever
+someone accesses this route a controller called `Album\Controller\List` will be accessed and more precisely its
+`indexAction()`. The problem is that this controller does not yet exist and if you reload the page you will be greeted
+with this lovely error message:
```
Comment (tfountain) [inline 01_writing_first_module.md]: This is also a bit wordy, how about:

> We've now created a route called `album-default` that listens to the URL `domain.loc/album`. Whenever someone accesses this route, the `indexAction()` function of the `Album\Controller\List` will be executed. However, this controller does not exist yet, so if you reload the page you will see this error message:

File: 01_writing_first_module.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    ),
+    'router' => array(
+        // our previous route configuration
+    )
+);
+```
+
+The above configuration lets the application know that the controller key `Album\Controller\List` is meant as an alias
+for the class `ListController` under the namespace `Album\Controller`. Sadly though this isn't enough yet. Reloading
+the page results into yet another error:
```
Comment (tfountain) [inline 01_writing_first_module.md]: Also a bit wordy, and don't be apologetic about the errors the user is seeing, as your approach is walking them through it in this way to help them learn what the errors mean and how to fix them ;)

How about:

> This configuration defines `Album\Controller\List` as an alias for the `ListController` under the namespace `Album\Controller`. Reloading the page should then give you:

File: 01_writing_first_module.md [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+```text
+( ! ) Fatal error: Class 'Album\Controller\ListController' not found in {libPath}/Zend/ServiceManager/AbstractPluginManager.php on line {lineNumber}
+```
+
+This error tells us that the application know what class we want to access but sadly it simply cannot find the class.
+That happens because our module never told the application where to find the classes that our module provides. Telling
+the application where to find classes is done by providing [autoloader](http://php.net/manual/de/language.oop5.autoload.php)
+configuration. This is done by addíng and implementing the [`AutoloaderProviderInterface`](https://github.com/zendframework/zf2/:current_branch/library/Zend/ModuleManager/Feature/AutoloaderProviderInterface.php)
+to our `Module` class. Note that once again adding the interface is optional, all you need is the `getAutoloaderConfig()`
+function defined by the interface.
```
Comment (tfountain) [inline 01_writing_first_module.md]: Also wordy:

> This error tells us that the application knows what class to load, but not where to find it. To fix this, we need to configure [autoloading](http://www.php.net/manual/en/language.oop5.autoload.php) for our Module. Autoloading is a process to allow PHP to automatically load classes on demand. For our Module we set this up by adding a `getAutoloaderConfig()` function to our Module class. (This function is defined in the  [`AutoloaderProviderInterface`](https://github.com/zendframework/zf2/:current_branch/library/Zend/ModuleManager/Feature/AutoloaderProviderInterface.php), although the presence of the function is enough, actually implementing the interface is optional.)

The PHP manual link should also point at the English language version.

File: 01_writing_first_module.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+controller itself. Let's create a controller class `Album\Controller\ListController`. To reflect [`PSR-0`](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-0.md)
+we do separate namespaces into folders so you'll find this file under
+`/module/Album/src/Album/Controller/ListController.php`:
+
+```php
+<?php
+// Filename: /module/Album/src/Album/Controller/ListController.php
+namespace Album\Controller;
+
+class ListController {}
```
Comment (tfountain) [inline 01_writing_first_module.md]: Same as with the module class definition, let's use 

```
class ListController
{

}
```

just to make it easier for users to edit later.

File: 01_writing_first_module.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        return include __DIR__ . '/config/module.config.php';
+    }
+}
+```
+
+Now this looks like a lot of change but don't be afraid. The [`AutoloaderProviderInterface`](https://github.com/zendframework/zf2/:current_branch/library/Zend/ModuleManager/Feature/AutoloaderProviderInterface.php)
+defines that our module has to tell the application where to find its classes. For this we have a pretty easy
+implementation of the `Zend\Loader\StandardAutoloader` that tells the application that all classes under the namespace
+`__NAMESPACE__` (`Album`) are found inside the `/src` folder inside the current directory and then once again inside a
+folder that's named like our namespace. So:
```
Comment (tfountain) [inline 01_writing_first_module.md]: Not sure if I'd mention the `AutoloaderProviderInterface` again here, as that was explained above. Perhaps:

> Now this looks like a lot of change but don't be afraid. We've added an `getAutoloaderConfig()` function which provides configuration for the `Zend\Loader\StandardAutoloader`. This configuration tells the application that classes in `__NAMESPACE__` (`Album`) can be found inside `__DIR__ . '/src/' . __NAMESPACE__` (`/module/Album/src/Album`).

File: 01_writing_first_module.md [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+All classes that have the namespace `Album` are found inside `/module/Album/src/Album`. This holds true as long as we
+hold on to [`PSR-0`](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-0.md). `PSR-0` is a community
+driven standard that describes the mandatory requirements that must be adhered to for autoloader interoperability.
```
Comment (tfountain) [inline 01_writing_first_module.md]: PSR-0 is brought up a bit out of the blue here. Let's tie its explanation to the StandardAutoloader:

> The `Zend\Loader\StandardAutoloader` uses a PHP community driven standard called [`PSR-0`](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-0.md). Amongst other things, this standard defines a way for PHP to map class names to the file system. So with this configured, the application knows that our `Album\Controller\ListController` class should exist at `/module/Album/src/Album/Controller/ListController.php`.

File: 01_writing_first_module.md [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+If you refresh the browser now you'll see that the error remains the same. And this is logical because remember what
+we did by now. We do have told the application where to find the controller now, but we have yet to actually write the
+controller itself. Let's create a controller class `Album\Controller\ListController`. To reflect [`PSR-0`](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-0.md)
+we do separate namespaces into folders so you'll find this file under
+`/module/Album/src/Album/Controller/ListController.php`:
```
Comment (tfountain) [inline 01_writing_first_module.md]: Some of this I put above, so maybe just:

> If you refresh the browser now you'll see the same error, as although we've configured the autoloader, we still need to create the controller class. Let's create this file now:

File: 01_writing_first_module.md [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+<?php
+// Filename: /module/Album/src/Album/Controller/ListController.php
+namespace Album\Controller;
+
+use Zend\Mvc\Controller\AbstractActionController;
+
+class ListController extends AbstractActionController {}
+```
+
+It's now time for another refresh of the site. Big surprise, there's another error message for you.
```
Comment (tfountain) [inline 01_writing_first_module.md]: This sounds a bit negative, perhaps just:

> It's now time for another refresh of the site. You should now see:

File: 01_writing_first_module.md (hunk 1 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: Comment incorporated into the next push in a few minutes. :thumbsup:

File: 01_writing_first_module.md (hunk 2 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup: can be used just like this!

File: 01_writing_first_module.md (hunk 3 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup:

File: 01_writing_first_module.md (hunk 4 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup: implemented.

File: 01_writing_first_module.md (hunk 6 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup:

File: 01_writing_first_module.md (hunk 7 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup:

File: 01_writing_first_module.md (hunk 8 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup:

File: 01_writing_first_module.md (hunk 5 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: :thumbsup: will be implemented for all future examples, too

File: 01_writing_first_module.md (hunk 9 above)
Comment (manuakasam) [inline 01_writing_first_module.md]: I wanted to sound funny :'(

:thumbsup:

File: 01_writing_first_module.md [hunk 10]
```diff
@@ -0,0 +1,354 @@
+Introducing our first "Album" Module
+====================================
+
+Now that we know about the basics of the Zend Framework 2 Skeleton Application, let's continue and create our very own
+module. We will create a module named "Album". This module will display a list of database entries that represent a
+single music album. Each album will receive a couple of properties like `id`, `artist` and `title`. We will create forms
```
Comment (GeeH) [inline 01_writing_first_module.md:6]: couple = two, suggest "number"

File: 01_writing_first_module.md [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+to enter new albums into our database and to edit existing albums. Furthermore we will do so by using best-practices
+throughout the whole QuickStart.
+
```
Comment (GeeH) [inline 01_writing_first_module.md:9]: I'd like to see intro rewritten, at the moment you are not saying you are doing anything different than the original quickstart - is this meant to be replacing or expanding on current album tutorial. I think you need to decide that and then tweak the intro.

File: 01_writing_first_module.md [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+Writing a new Module
+====================
+
+Let's start by creating a new folder under the `/module` directory called `Album`. Your directory structure should look
+like this:
+
+[[ DISPLAY IMAGE - DIRECTORY STRUCTURE /module/Album ]]
+
+To be recognized as a module by the [ModuleManager](https://github.com/zendframework/zf2/:current_branch/library/Zend/ModuleManager/ModuleManager.php)
+all we need to do is create a PHP-Class named `Module` under our module's namespace, which is `Album`. Create the file
```
Comment (GeeH) [inline 01_writing_first_module.md:20]: PHP-Class = PHP class (not hyphenated)

File: 01_writing_first_module.md [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+interacting with a third party API.
+
+Organizing your code into modules makes it easier for you to reuse functionality in other application, or to use
+modules written by the community.
+
+Configuring the Module
+======================
+
+The next thing we're going to do is add a route to our application so that our module can be accessed through the
+URL `localhost:8080/album`. We do this by adding router configuration to our module, but first we need to let the
```
Comment (GeeH) [inline 01_writing_first_module.md:66]: How do you know which url and port they are using here? Have I missed a previous chapter that explains installing skeleton app and running it under that url?

File: 01_writing_first_module.md [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+class Module implements ConfigProviderInterface
+{
+    /**
+     * Returns configuration to merge with application configuration
+     *
+     * @return array|\Traversable
+     */
+    public function getConfig()
+    {
+        return array();
```
Comment (GeeH) [inline 01_writing_first_module.md:90]: CONTENTION ALERT!!!

:warning: :warning: :warning: :warning: :warning: 

Have you considered using php 5.4 array syntax for readability. Probably a stupid idea.

File: 01_writing_first_module.md [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+<?php
+// Filename: /module/Album/src/Album/Controller/ListController.php
+namespace Album\Controller;
+
+class ListController
+{
+}
+```
+
+Reloading the page now will finally result into a new screen. The new error message looks like this:
```
Comment (GeeH) [inline 01_writing_first_module.md:262]: :question: "result into"

File: 01_writing_first_module.md [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+Zend\Mvc\Exception\InvalidControllerException
+
+File:
+{libraryPath}/Zend/Mvc/Controller/ControllerManager.php:{lineNumber}
+Message:
+Controller of type Album\Controller\ListController is invalid; must implement Zend\Stdlib\DispatchableInterface
+```
+
+This happens because our controller must implement [`Zend\Stdlib\DispatchableInterface`](https://github.com/zendframework/zf2/:current_branch/library/Zend/Stdlib/DispatchableInterface.php) in order to be 'dispatched'
+(or run) by ZendFramework's MVC layer. ZendFramework provides some base controller implementation of it with
```
Comment (GeeH) [inline 01_writing_first_module.md:282]: Product is called Zend Framework 2 or ZF2 - this won't get accepted by Zend representative without the product name being consistent.

File: 01_writing_first_module.md [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+Now the application tells you that a view template-file can not be rendered. Given our current progress this is more
+than natural, because we have yet to actually write this view-file ourselves. The standard path would be
+`/module/Album/view/album/list/index.phtml`. Create this file and add some dummy content to it:
+
+```html
+<!-- Filename: /module/Album/view/album/list/index.phtml -->
+<h1>Album\ListController::indexAction()</h1>
+```
+
+Before we continue let us quickly take a look at where we placed this file. First off, view files are not to be found
```
Comment (GeeH) [inline 01_writing_first_module.md:322]: First off = Firstly

File: 01_writing_first_module.md [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+under the `/src` directory because they are not source files. They are views so `/view` is much more logical. The
```
Comment (GeeH) [inline 01_writing_first_module.md:323]: they are not PHP classes which is why they aren't found under the PSR-0 root - I think the term "source files" is confusing

File: 01_writing_first_module.md [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+succeeding path however deserves some explanation but it's very simple. First we have the lowercased namespace. Following
```
Comment (GeeH) [inline 01_writing_first_module.md:324]: ? "succeeding"

Comment (GeeH) [inline 01_writing_first_module.md:324]: Following = Followed

File: 01_writing_first_module.md [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+    'router'      => array( /** Route Configuration */ )
+);
+```
+
+The above configuration tells the application that the folder `/module/Album/view` has view files in it that match the
+above described default scheme. It is important to note that with this you can not only ship view files for your module
+but you can also overwrite view files from other modules. You can find more information about this inside our Appendix
+section of the QuickStart. //@todo Appendix link
+
+Reload your site now. Finally we are at a point where we see something different than an error display. Congratulations,
```
Comment (GeeH) [inline 01_writing_first_module.md:351]: display = being displayed

File: 01_writing_first_module.md [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+not only have you created a simple "Hello World" kinda module, you also learned about many error messages and their
```
Comment (GeeH) [inline 01_writing_first_module.md:352]: "kinda" feels out of place here, you haven't used slang or this kind of speak anywhere else in the chapter, suggest "style"

File: 01_writing_first_module.md [hunk 22]
```diff
@@ -0,0 +1,354 @@
...
+sources. If we didn't exhaust you too much, continue with our QuickStart and let's create a module that actually does
```
Comment (GeeH) [inline 01_writing_first_module.md:353]: sources = causes

File: 01_writing_first_module.md (hunk 14 above)
Comment (akrabat) [inline 01_writing_first_module.md:90]: I disagree with @GeeH - ZF2 is PHP 5.3+, so the Quick Start should work on PHP 5.3 and short array syntax isn't important enough to worry about a call out for.

File: 01_writing_first_module.md (hunk 11 above)
Comment (manuakasam) [inline 01_writing_first_module.md:9]: In fact a preample is still to be written. I wanted to do a full installation requirements chapter but that's really covered already. It'll be done sometime tomorrow. :thumbsup:

File: 01_writing_first_module.md (hunk 14 above)
Comment (GeeH) [inline 01_writing_first_module.md:90]: Point taken and agreed with. I'll leave this note here so everyone can laugh at my stupid idea.

File: 01_writing_first_module.md (hunk 10 above)
Comment (manuakasam) [inline 01_writing_first_module.md:6]: :thumbsup:

File: 01_writing_first_module.md (hunk 13 above)
Comment (manuakasam) [inline 01_writing_first_module.md:66]: Another point for the preamble chapter. I stil suggest using `localhost:8080` throughout the tutorial though it will be noted that this could be replaced by any vhost set up by the user.

File: 01_writing_first_module.md (hunk 14 above)
Comment (manuakasam) [inline 01_writing_first_module.md:90]: :thumbsdown: Considering this tutorial will probably go live during ZF 2.x we're restricted to 5.3 syntax

File: 01_writing_first_module.md (hunk 17 above)
Comment (akrabat) [inline 01_writing_first_module.md:322]: First off, => Note that. Also be positive, not the negative. Something like "Note that view files are found the view subdirectory, not src as they are no PHP class files, but template files for rendering HMTL."

File: 01_writing_first_module.md (hunk 15 above)
Comment (manuakasam) [inline 01_writing_first_module.md:262]: hmm, "reveal a new screen", "show a new screen"?

File: 01_writing_first_module.md (hunk 16 above)
Comment (manuakasam) [inline 01_writing_first_module.md:282]: I'd vote for ZF2.

Comment (GeeH) [inline 01_writing_first_module.md:282]: Just a point of note that Zend marketing types need the products to be referenced correctly and ZendFramework is not the product name :)

File: 01_writing_first_module.md (hunk 17 above)
Comment (manuakasam) [inline 01_writing_first_module.md:322]: :thumbsup: for akrabats formulation.

File: 01_writing_first_module.md (hunk 16 above)
Comment (akrabat) [inline 01_writing_first_module.md:282]: I would use "the MVC layer" :)

File: 01_writing_first_module.md [hunk 23]
```diff
@@ -0,0 +1,354 @@
...
+
+Configuring the Module
+======================
+
+The next thing we're going to do is add a route to our application so that our module can be accessed through the
+URL `localhost:8080/album`. We do this by adding router configuration to our module, but first we need to let the
+`ModuleManager` know that our module has configuration that it needs to load.
+
+This is done by adding a `getConfig()` function to the `Module` class that returns the configuration. (This function is
+defined in the `ConfigProviderInterface` although actually implementing this interface in the module class is optional.)
```
Comment (haclong) [inline 01_writing_first_module.md:70]: something not clear here : the ConfigProviderInterface is optional..; than means we can either do 
class Module
{
    public function getconfig()
}

or 
class Module implements ConfigProviderInterface
{
    public function getConfig()
}

but what is confusing is, - maybe - that both writings are ok...
i'd think that if something is optional, it means you can still use your Module class but if you need to access some extra methods (such as getConfig()), you'd have to implements the ConfigProviderInterface.

But in ZF2, actually, you can access getConfig while implementing or not the interface...
(same remarks for the serviceManager interface later)
